src/wallet/migrate.cpp

```diff
@@ -0,0 +1,771 @@
+        db_file >> file >> offset;
+        if (file != 0 || offset != 1) {
+            throw std::runtime_error("LSNs are not reset, this database is not completely flushed. Please reopen then close the database with a version that has BDB support");
+        }
+    }
+
+    // Read the root page
+    SeekToPage(db_file, outer_meta.root, page_size);
+    PageHeader header(outer_meta.root, outer_meta.other_endian);
+    db_file >> header;
+    if (header.type != PageType::BTREE_LEAF) {
+        throw std::runtime_error("Unexpected outer database root page type");
+    }
+    if (header.entries != 2) {
+        throw std::runtime_error("Unexpected number of entries in outer database root page");
+    }
+    RecordsPage page(header);
+    db_file >> page;
+
+    // First record should be the string "main"
+    if (!std::holds_alternative<DataRecord>(page.records.at(0)) || std::get<DataRecord>(page.records.at(0)).data != SUBDATABASE_NAME) {
+        throw std::runtime_error("Subdatabase has an unexpected name");
+    }
+    // Check length of page number for subdatabase location
+    if (!std::holds_alternative<DataRecord>(page.records.at(1)) || std::get<DataRecord>(page.records.at(1)).m_header.len != 4) {
+        throw std::runtime_error("Subdatabase page number has unexpected length");
+    }
+
+    // Read subdatabase page number
+    // It is written as a big endian 32 bit number
+    uint32_t main_db_page = ReadBE32(UCharCast(std::get<DataRecord>(page.records.at(1)).data.data()));
+
+    // The main database is in a page that doesn't exist
+    if (main_db_page > outer_meta.last_page) {
+        throw std::runtime_error("Page number is greater than database last page");
+    }
+
+    // Read the inner metapage
+    SeekToPage(db_file, main_db_page, page_size);
+    MetaPage inner_meta(main_db_page);
+    db_file >> inner_meta;
+
+    if (inner_meta.pagesize != page_size) {
+        throw std::runtime_error("Unexpected page size");
+    }
+
+    if (inner_meta.last_page > outer_meta.last_page) {
+        throw std::runtime_error("Subdatabase last page is greater than database last page");
+    }
+
+    // Make sure encryption is disabled
+    if (inner_meta.encrypt_algo != 0) {
+        throw std::runtime_error("BDB builtin encryption is not supported");
+    }
+
+    // Do a DFS through the BTree, starting at root
+    std::vector<uint32_t> pages{inner_meta.root};
+    while (pages.size() > 0) {
+        uint32_t curr_page = pages.back();
+        if (curr_page > inner_meta.last_page) {
+            std::runtime_error("Page number is greater than subdatabase last page");
+        }
+        pages.pop_back();
+        SeekToPage(db_file, curr_page, page_size);
+        PageHeader header(curr_page, inner_meta.other_endian);
+        db_file >> header;
+        if (header.type != PageType::BTREE_LEAF && header.type != PageType::BTREE_INTERNAL) {
+            throw std::runtime_error("Unexpected inner database page type");
```

Comment:
`header.type` is compared against `BTREE_LEAF` and `BTREE_INTERNAL`, which are the same cases as handled in the `switch()` statement below, which also errors in the default page (though with a slightly different error), so this check is redundant?

Reply:
Good point. Removed this check.